### ISSUE TITLE
CAPV-generated Cert only includes first NIC address

### ISSUE THREAD

**Issue by MaxRink** [kubernetes-sigs/cluster-api-provider-vsphere#663]
/kind bug

**What steps did you take and what happened:**
Im trying to get CAPV up and running with multiple NICs due to routing limitations of our Networks.
Ive assigend two NICs in the Controlplane.yml and selected one as preffered CIDR for the API

``` yml

 network:
    preferredAPIServerCidr: "10.88.168.0/24"
    devices:
      - networkName: VLAN339
        ipAddrs: 
          - "10.27.50.11/24"
        gateway4: "10.27.50.1"
        nameservers: 
          - "8.8.8.8"
        dhcp4: false
        dhcp6: false
      - networkName: VLAN340
        ipAddrs: 
          - "10.88.168.11/24"
        gateway4: "10.88.168.1"
        dhcp4: false
        dhcp6: false   

```

When i then boostrap the Cluster the boostrapping fails with this logline:
```E1118 16:00:22.678808       1 controller.go:218] controller-runtime/controller "msg"="Reconciler error" "error"="failed to reconcile cloud config secret for VSphereCluster default/management-cluster: failed to create cloud provider secret for Cluster default/management-cluster: Post https://10.88.168.11:6443/api/v1/namespaces/kube-system/secrets: x509: certificate is valid for 100.64.0.1, 10.27.50.11, not 10.88.168.11"  "controller"="vspherecluster" "request"={"Namespace":"default","Name":"management-cluster"}```

So the API cert got issued for the wrong IP
**What did you expect to happen:**
The API Cert gets issued for the address within the specified "preferred CIDR"

**Environment:**

- Cluster-api-provider-vsphere version: 0.5.3
- Kubernetes version: (use `kubectl version`): 1.16.2
- OS (e.g. from `/etc/os-release`):  Ubuntu

**Comment by akutz** [comment]
Hi @MaxRink,

I think there may be some confusion as to the intended use of the `preferredAPIServerCidr` field. The purpose of the field is not to influence which IP is selected by the guest operating system, but to indicate to CAPV which IP *will be* selected by the guest operating system.

The use case we considered when creating the notion of the `preferredAPIServerCidr` is related to DHCP. For example:

* A machine is defined with dual network interfaces
* Both network interface uses DHCP
* Both network interfaces are connected to networks whose DHCP servers are configured to advertise a gateway for IPs leased from the segment (please see the [`option routers` configuration from dhcpd-options](https://linux.die.net/man/5/dhcpd-options) for an example)

Now, your configuration with static addresses is not far from the above scenario. The thing is, whether by DHCP or statically, having multiple gateways is a bad time. It makes it impossible to deterministically discover the _local_ IP address. What does that mean? Well, the [IP discovery code](https://github.com/vmware/cloud-init-vmware-guestinfo/blob/cb9e8e4790acacb06b1f473565535f6c6f4a595c/DataSourceVMwareGuestInfo.py#L342-L413) is in the cloud-init datasource built into the machine images. The IP discovery workflow looks something like this:

1. Get the _default_ gateway

Full stop. We don't even have to go past step one to find the problem. Guests with multiple gateways defined mean there's no clear means of determining the *default* gateway. Without that, there's no way to find the IPv4 or IPv6 address on the same device as the gateway. And thus we cannot determine the default, local IP address.

So let's alter the original scenario some:

* A machine is defined with dual network interfaces
* Both network interface uses DHCP
* One of the network interfaces is connected to networks whose DHCP server is configured to advertise a gateway for IPs leased from the segment
* One of the network interfaces is connected to a network whose DHCP server is **not** configured to advertise a gateway for IPs leased from the segment

On boot there **will** be a default gateway and thus the cloud-init datasource **can** figure out the default, local IP address. This is the address used by kubeadm and for which certificates are generated. 

However, there's still an issue. When we inspect the IP addresses of the VM from outside the guest OS, there's no way to know which of the two IP addresses is the default IP address external to the guest!

Enter the notion of the `preferredAPIServerCidr`. Here's [the utility function](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/c45a65d65a2b4d124defe2c90cee492f8693588d/pkg/util/machines.go#L106-L130) that leverages the field:

```golang
// GetMachinePreferredIPAddress returns the preferred IP address for a
// VSphereMachine resource.
func GetMachinePreferredIPAddress(machine *infrav1.VSphereMachine) (string, error) {
	var cidr *net.IPNet
	if cidrString := machine.Spec.Network.PreferredAPIServerCIDR; cidrString != "" {
		var err error
		if _, cidr, err = net.ParseCIDR(cidrString); err != nil {
			return "", errors.New("error parsing preferred API server CIDR")
		}
	}

	for _, nodeAddr := range machine.Status.Addresses {
		if nodeAddr.Type != corev1.NodeInternalIP {
			continue
		}
		if cidr == nil {
			return nodeAddr.Address, nil
		}
		if cidr.Contains(net.ParseIP(nodeAddr.Address)) {
			return nodeAddr.Address, nil
		}
	}

	return "", ErrNoMachineIPAddr
}
```

Because a machine may have several IP addresses, the above function iterates over the list and selects the one that falls into the CIDR specified by `preferredAPIServerCidr` as the IP address that:

* Cloud-init recorded as the local IP address
* and Kubeadm used as the API server address

So as you can see, the `preferredAPIServerCidr` field is meant to disambiguate a list of IP addresses returned from inspecting a VM externally. There needs to be a way to indicate which IP is the one that the machine would have used to initialize Kubernetes. When DHCP is in use, there's no way to know the IPs ahead of time, so a CIDR that can be used by an operator to say "I don't know why my IP will be, but I know it will belong to this network segment" is the best solution we have.

Your case is even more interesting:

* Multiple devices
* Each with a static IP
* Each with a defined gateway

What is astounding to me is that the cloud-init datasource determined a local IP address at all given that multiple gateways tends to [confuse the library](https://github.com/al45tair/netifaces) used by the IP discovery logic:

> BTW, if you're trying to configure your machine to have multiple gateways for the same address family, it's a very good idea to check the documentation for your operating system very carefully, as some systems become extremely confused or route packets in a non-obvious manner.
> 
> I'm very interested in hearing from anyone (on any platform) for whom the gateways() method doesn't produce the expected results. It's quite complicated extracting this information from the operating system (whichever operating system we're talking about), and so I expect there's at least one system out there where this just won't work.

Here's my advice:

* Drop the `preferredAPIServerCidr` as it serves no purpose in your case
* Reorder the network devices so `VLAN340` is the first device

It seems with multiple network devices using static addressing, the first device is selected as the local IP address. I can't say this is deterministic, and honestly I'm surprised the datasource didn't balk at seeing two gateways.

FWIW, CAPV **does** support defining multiple, static routes. If you **need** the second device's traffic to go to a specific gateway, I'd suggest using a static route to achieve that.

Bottom line -- unless you override the generated `KubeadmConfig` resources to use a lookup pattern other than `ds.meta_data.local-ipv4` to discover the name of the local IP address, then deployed systems will continue to use the default IP address for initializing the API server.

**Comment by akutz** [comment]
ping @MaxRink

**Comment by MaxRink** [comment]
Ok, understood.
THX for the detailed explanation.
And sorry for the late answer, i was on holidays until today.
